nemo/collections/tts/modules/submodules.py

```diff
@@ -410,3 +419,284 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
+    def input_types(self):
+        return {
+            "inp": NeuralType(('B', 'D', 'T_spec'), MelSpectrogramType()),
+            "inp_mask": NeuralType(('B', 'T_spec', 1), TokenDurationType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "gst": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inp, inp_mask):
+        style_embedding = self.reference_encoder(inp, inp_mask)
+        gst = self.style_attention(style_embedding)
+        return gst
+
+
+class ReferenceEncoder(NeuralModule):
+    """
+    Encode mel-spectrograms to an utterance level feature
+    """
+
+    def __init__(self, cnn_filters=[32, 32, 64, 64, 128, 128], dropout=0.2, gru_hidden=128):
+        super(ReferenceEncoder, self).__init__()
+        self.filter_size = [1] + cnn_filters
+        self.dropout = dropout
+        self.conv = torch.nn.Sequential(
+            OrderedDict(
+                [
+                    module
+                    for i in range(len(cnn_filters))
+                    for module in (
+                        ("transpose_bchw_{}".format(i + 1), Transpose([0, 3, 1, 2])),
+                        (
+                            "conv2d_{}".format(i + 1),
+                            torch.nn.Conv2d(
+                                in_channels=int(self.filter_size[i]),
+                                out_channels=int(self.filter_size[i + 1]),
+                                kernel_size=(3, 3),
+                                stride=(2, 2),
+                                padding=(1, 1),
+                                dilation=(1, 1),
+                                bias=True,
+                            ),
+                        ),
+                        ("transpose_bhwc_{}".format(i + 1), Transpose([0, 2, 3, 1])),
+                        ("relu_{}".format(i + 1), torch.nn.ReLU()),
+                        ("layer_norm_{}".format(i + 1), torch.nn.LayerNorm(self.filter_size[i + 1]),),
+                        ("dropout_{}".format(i + 1), torch.nn.Dropout(self.dropout)),
+                    )
+                ]
+            )
+        )
+
+        self.gru = torch.nn.GRU(input_size=cnn_filters[-1] * 2, hidden_size=gru_hidden, batch_first=True,)
+
+    @property
+    def input_types(self):
+        return {
+            "inputs": NeuralType(('B', 'D', 'T_spec'), MelSpectrogramType()),
+            "inputs_masks": NeuralType(('B', 'T_spec', 1), TokenDurationType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "out": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inputs, inputs_masks):
+        inputs = inputs.transpose(1, 2)
+
+        inputs = inputs * inputs_masks
+        out = inputs.unsqueeze(3)
+        out = self.conv(out)
+        out = out.view(out.shape[0], out.shape[1], -1).contiguous()
+        self.gru.flatten_parameters()
+        _, out = self.gru(out)
+
```

Comment:
I had this comment on the previous adapter CR last year:

Masking is required in between CNN layers (In theory, maybe not always in practice. Definitely required for final GRU layer). This is relatively complicated with reference encoder due to (2, 2) stride in CNN cutting sequence lengths in half at each step.

I suggest copying implementation from this or other github repo, use it to create mask after each CNN layer, and ideally adding a unit test to it: https://github.com/mozilla/TTS/blob/e9e07844b77a43fb0864354791fb4cf72ffded11/TTS/tts/layers/gst_layers.py#L84

Reply:
Main concern is not masking RNN input is a large bug likely to impact performance, and relatively easy to fix. Masking in between each CNN is less important and more difficult to implement.

Reply:
Masks between CNN layers and before RNN are added